File: receiver-windows/src/OpenAudioLink/Protocol/PacketWriter.cs
using System;

namespace OpenAudioLink.Protocol
{
    public static class PacketWriter
    {
        public static byte[] WritePacket(byte packetType, uint sequenceNumber, ulong timestamp, byte[] payload)
        {
            payload = payload ?? new byte[0];
            if (payload.Length > ProtocolConstants.MaxPacketSize)
            {
                throw new ArgumentOutOfRangeException(nameof(payload), "Payload is too large.");
            }

            byte[] packet = new byte[ProtocolConstants.HeaderSize + payload.Length];
            Buffer.BlockCopy(ProtocolConstants.Magic, 0, packet, 0, ProtocolConstants.Magic.Length);
            packet[4] = ProtocolConstants.MajorVersion;
            packet[5] = ProtocolConstants.MinorVersion;
            packet[6] = packetType;
            packet[7] = 0;
            WriteUInt32(packet, 8, sequenceNumber);
            WriteUInt64(packet, 12, timestamp);
            WriteUInt32(packet, 20, (uint)payload.Length);
            Buffer.BlockCopy(payload, 0, packet, ProtocolConstants.HeaderSize, payload.Length);
            return packet;
        }

        internal static void WriteUInt16(byte[] buffer, int offset, ushort value)
        {
            buffer[offset] = (byte)(value >> 8);
            buffer[offset + 1] = (byte)value;
        }

        internal static void WriteUInt32(byte[] buffer, int offset, uint value)
        {
            buffer[offset] = (byte)(value >> 24);
            buffer[offset + 1] = (byte)(value >> 16);
            buffer[offset + 2] = (byte)(value >> 8);
            buffer[offset + 3] = (byte)value;
        }

        internal static void WriteUInt64(byte[] buffer, int offset, ulong value)
        {
            WriteUInt32(buffer, offset, (uint)(value >> 32));
            WriteUInt32(buffer, offset + 4, (uint)value);
        }
    }
}

File: sender-android/app/src/main/java/com/openaudiolink/protocol/PacketWriter.kt
package com.openaudiolink.protocol

object PacketWriter {
    fun write(packetType: Int, sequenceNumber: Long, timestamp: Long, payload: ByteArray = byteArrayOf()): ByteArray =
        ByteArray(ProtocolConstants.HeaderSize + payload.size).also { packet ->
            ProtocolConstants.Magic.copyInto(packet)
            packet[4] = ProtocolConstants.MajorVersion.toByte()
            packet[5] = ProtocolConstants.MinorVersion.toByte()
            packet[6] = packetType.toByte()
            packet[7] = 0
            writeUInt32(packet, 8, sequenceNumber)
            writeUInt64(packet, 12, timestamp)
            writeUInt32(packet, 20, payload.size.toLong())
            payload.copyInto(packet, ProtocolConstants.HeaderSize)
        }

    internal fun writeUInt16(buffer: ByteArray, offset: Int, value: Int) {
        buffer[offset] = (value ushr 8).toByte()
        buffer[offset + 1] = value.toByte()
    }

    internal fun writeUInt32(buffer: ByteArray, offset: Int, value: Long) {
        buffer[offset] = (value ushr 24).toByte()
        buffer[offset + 1] = (value ushr 16).toByte()
        buffer[offset + 2] = (value ushr 8).toByte()
        buffer[offset + 3] = value.toByte()
    }

    internal fun writeUInt64(buffer: ByteArray, offset: Int, value: Long) {
        writeUInt32(buffer, offset, value ushr 32)
        writeUInt32(buffer, offset + 4, value)
    }
}
